test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/TempDataApplicationModelProviderTest.cs

```diff
@@ -0,0 +1,132 @@
+
+            var context = new ApplicationModelProviderContext(new[] { typeof(TestController_OneTempDataProperty).GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act
+            provider.OnProvidersExecuting(context);
+
+            // Assert
+            var controller = Assert.Single(context.Result.Controllers);
+            Assert.Single(controller.Filters, f => f is SaveTempDataPropertyFilterFactory);
+        }
+
+        [Fact]
+        public void CreateControllerModelWithTwoTempDataProperties_TempDataAttributeAddsTempDataFilter()
+        {
+            // Arrange
+            var provider = new TempDataApplicationModelProvider();
+            var defaultProvider = new DefaultApplicationModelProvider(new TestOptionsManager<MvcOptions>());
+
+            var context = new ApplicationModelProviderContext(new[] { typeof(TestController_TwoTempDataProperties).GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act
+            provider.OnProvidersExecuting(context);
+
+            // Assert
+            var controller = Assert.Single(context.Result.Controllers);
+            Assert.Single(controller.Filters, f => f is SaveTempDataPropertyFilterFactory);
+        }
+
+        [Fact]
+        public void ThrowsInvalidOperationException_PrivateSetter()
+        {
+            // Arrange
+            var provider = new TempDataApplicationModelProvider();
+            var defaultProvider = new DefaultApplicationModelProvider(new TestOptionsManager<MvcOptions>());
+
+            var context = new ApplicationModelProviderContext(new[] { typeof(TestController_PrivateSet).GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                provider.OnProvidersExecuting(context);
+            });
+
+            Assert.Equal("TempData property Test does not have a public getter or setter.", exception.Message);
+        }
+
+        [Fact]
+        public void ThrowsInvalidOperationException_NonPrimitiveType()
+        {
+            // Arrange
+            var provider = new TempDataApplicationModelProvider();
+            var defaultProvider = new DefaultApplicationModelProvider(new TestOptionsManager<MvcOptions>());
+
+            var context = new ApplicationModelProviderContext(new[] { typeof(TestController_NonPrimitiveType).GetTypeInfo() });
+            defaultProvider.OnProvidersExecuting(context);
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                provider.OnProvidersExecuting(context);
+            });
+
+            Assert.Equal("TempData property Test is not declared as a primitive type or string.", exception.Message);
+        }
+
+
```

Comment:
Remove extra line. Also needs tests to verify that it initializes the filter with the right PropertyHelpers